src/metabase/db/setup.clj

```diff
@@ -125,37 +152,47 @@
   entries for the \"magic\" groups and permissions entries. "
   false)
 
-(s/defn ^:private run-schema-migrations!
+(mu/defn ^:private run-schema-migrations!
   "Run through our DB migration process and make sure DB is fully prepared"
-  [db-type       :- s/Keyword
-   data-source   :- javax.sql.DataSource
-   auto-migrate? :- (s/maybe s/Bool)]
+  [db-type       :- :keyword
+   data-source   :- (ms/InstanceOfClass javax.sql.DataSource)
+   auto-migrate? :- [:maybe :boolean]]
   (log/info (trs "Running Database Migrations..."))
   (migrate! db-type data-source (if auto-migrate? :up :print))
   (log/info (trs "Database Migrations Current ... ") (u/emoji "✅")))
 
-(s/defn ^:private run-data-migrations!
+(mu/defn ^:private run-data-migrations!
   "Do any custom code-based migrations now that the db structure is up to date."
   []
   ;; TODO -- check whether we can remove the circular ref busting here.
   (when-not *disable-data-migrations*
     (classloader/require 'metabase.db.data-migrations)
     ((resolve 'metabase.db.data-migrations/run-all!))))
 
+(defn- fresh-install?
+  [db-type data-source]
+  (not (liquibase/table-exists? (if (= :h2 db-type)
+                                  "CORE_USER"
+                                  "core_user") data-source)))
+
 ;; TODO -- consider renaming to something like `verify-connection-and-migrate!`
 ;;
 ;; TODO -- consider whether this should be done automatically the first time someone calls `getConnection`
-(s/defn setup-db!
+(mu/defn setup-db!
   "Connects to db and runs migrations. Don't use this directly, unless you know what you're doing;
   use [[metabase.db/setup-db!]] instead, which can be called more than once without issue and is thread-safe."
-  [db-type       :- s/Keyword
-   data-source   :- javax.sql.DataSource
-   auto-migrate? :- (s/maybe s/Bool)]
+  [db-type       :- :keyword
+   data-source   :- (ms/InstanceOfClass javax.sql.DataSource)
+   auto-migrate? :- [:maybe :boolean]]
   (u/profile (trs "Database setup")
     (u/with-us-locale
        (binding [mdb.connection/*application-db* (mdb.connection/application-db db-type data-source :create-pool? false) ; should already be a pool
                  setting/*disable-cache*         true]
-         (verify-db-connection   db-type data-source)
+         (verify-db-connection db-type data-source)
+         (when (fresh-install? db-type data-source)
```

Comment:
TODO: we might run into a race condition for this if someone start metabase as multi-instances